cellmlmanip/model.py

```diff
@@ -0,0 +1,508 @@
+            if dummy_number == expr:
+                return number_info['cellml:units']
+
+        # Note unit information about this expression in this component. It might be referring to
+        # a variable in a different component
+        variable_with_assigned = self.model.find_variable({'assignment': expr})
+        if variable_with_assigned:
+            return variable_with_assigned[0]['units']
+
+        # No information found about this
+        return None
+
+    def find_variable(self, search_dict):
+        """Finds variables in this model that match the given key: value pairs
+
+        :param search_dict:
+        :return: a dictionary of key: value pairs
+        """
+        # a list to collect all matched variables
+        matches = []
+
+        # for each defined variable in the component
+        for variable in self.variables.values():
+            # Every search (key, value) pair needs to be in the variable's attributes
+            matched = True
+            for search_key, search_value in search_dict.items():
+                # If the key is not in the variable or the value is different
+                if search_key not in variable or search_value != variable[search_key]:
+                    # This variable doesn't match, break
+                    matched = False
+                    break
+            if matched:
+                # All search (key: value)s were found in this variable
+                matches.append(variable)
+        return matches
+
+
+class Model(object):
+    """Holds all information about a CellML model and exposes it for manipulation (intention!)
+    """
+    def __init__(self, name: str) -> None:
+        """Create a new instance of Model object
+        :param name: the name of the model e.g. from <model name="">
+        """
+        self.name: str = name
+        self.units: 'QuantityStore' = None
+        self.components: Dict[str, Component] = {}
+        self.connections: List[Tuple] = []
+        self.rdf: rdflib.Graph = rdflib.Graph()
+
+    def __str__(self):
+        """Pretty-print each of the components in this model"""
+        return '\n'.join([str(v) for k, v in self.components.items()])
+
+    def add_unit(self, units_elements: dict):
+        """Adds information about <units> in <model>
+        """
+        self.units = QuantityStore(units_elements)
+
+    def add_component(self, component: Component):
+        """Adds name to list of <component>s in the <model>
+        """
+        self.components[component.name] = component
+
+    def add_rdf(self, rdf: str):
+        """ Takes RDF string and stores it in an RDFlib.Graph for the model. Can be called
+        repeatedly.
+        """
+        self.rdf.parse(StringIO(rdf))
+
+    def _is_free_variable(self, variables: Dict):
+        """Checks whether a variable gets its value for another component. There are two
+        possibilities. Either is has:
+            (i) public_interface="out" or
```

Comment:
Does a `public=out private=in` variable allow it to have a connection with another variable from within the same component?

Reply:
Not if I understand the question correctly. So say you have components A-E where A is the parent of B & C, and B is the parent of D & E, with variables as follows:
- D.v public=out
- E.v public=in connects to D.v
- B.v private=in public=out connects to D.v and A.v and C.v
- C.v public=in connects to B.v
- A.v private=in connects to B.v

In other words, when you have `public=out private=in` you don't need to export the variable privately (and indeed can't) since your child components will be siblings of the original exporter and can get the variable directly from there.

Reply:
I'm handling group encapsulated relationships now, and have updated the __connect() method accordingly. Could you check to see if I've got the logic right?